src/uploaders.rs

```diff
@@ -224,3 +228,22 @@ impl Drop for Bomb {
         }
     }
 }
+
+fn verify_tarball(krate: &Crate, vers: &semver::Version, tarball: &[u8]) -> CargoResult<()> {
+    let decoder = GzDecoder::new(tarball)?;
+    let mut archive = tar::Archive::new(decoder);
+    let prefix = format!("{}-{}", krate.name, vers);
+    for entry in archive.entries()? {
+        let entry = entry?;
+
+        // Verify that all entries actually start with `$name-$vers/`.
+        // Historically Cargo didn't verify this on extraction so you could
+        // upload a tarball that contains both `foo-0.1.0/` source code as well
+        // as `bar-0.1.0/` source code, and this could overwrite other crates in
+        // the registry!
+        if !entry.path()?.starts_with(&prefix) {
```

Comment:
Do you need a trailing '/', or is this looking at a whole directory entry?

Reply:
Can you add checks if any of the entries are: '..' or non-file/directories?

Reply:
The trailing `/` isn't necessary due to the way `Path` works, entries like `..` are tested in `tar` itself, and non-file/directory entries are a feature, not a bug, but have extra handling in `tar`.